stream_alert/rule_processor/handler.py

```diff
@@ -91,60 +122,35 @@ def get_alerts(self):
         """
         return self._alerts
 
-    def _kinesis_process(self, payload, classifier):
-        """Process Kinesis data for alerts"""
-        data = StreamPreParsers.pre_parse_kinesis(payload.raw_record)
-        self._process_alerts(classifier, payload, data)
-
-    def _s3_process(self, payload, classifier):
-        """Process S3 data for alerts"""
-        s3_file, s3_object_size = StreamPreParsers.pre_parse_s3(payload.raw_record)
-        count, processed_size = 0, 0
-        for data in StreamPreParsers.read_s3_file(s3_file):
-            payload.refresh_record(data)
-            self._process_alerts(classifier, payload, data)
-            # Add the current data to the total processed size, +1 to account for line feed
-            processed_size += (len(data) + 1)
-            count += 1
-            # Log an info message on every 100 lines processed
-            if count % 100 == 0:
-                avg_record_size = ((processed_size - 1) / count)
-                approx_record_count = s3_object_size / avg_record_size
-                LOGGER.info('Processed %s records out of an approximate total of %s '
-                            '(average record size: %s bytes, total size: %s bytes)',
-                            count, approx_record_count, avg_record_size, s3_object_size)
-
-    def _sns_process(self, payload, classifier):
-        """Process SNS data for alerts"""
-        data = StreamPreParsers.pre_parse_sns(payload.raw_record)
-        self._process_alerts(classifier, payload, data)
-
-    def _process_alerts(self, classifier, payload, data):
+    def _process_alerts(self, payload):
         """Process records for alerts and send them to the correct places
 
         Args:
-            classifier [StreamClassifier]: Handler for classifying a record's data
             payload [StreamPayload]: StreamAlert payload object being processed
-            data [string]: Pre parsed data string from a raw_event to be parsed
         """
-        classifier.classify_record(payload, data)
-        if not payload.valid:
-            if self.env['lambda_alias'] != 'development':
-                LOGGER.error('Record does not match any defined schemas: %s\n%s', payload, data)
-            self._failed_record_count += 1
-            return
+        for record in payload.pre_parse():
+            self.classifier.classify_record(record)
+            if not record.valid:
+                if self.env['lambda_alias'] != 'development':
+                    LOGGER.error('Record does not match any defined schemas: %s\n%s',
```

Comment:
Because it will get lost in the shuffle when printed in-line with testing output. The variable below (`self._failed_record_count += 1`) is what is used by the testing framework to handle error reporting for invalid records/logs. Basically, a similar messages gets printed when running in a development env, just not at this time and at the end of testing instead.

The alternative to this would be having a list of errors that get appended to, and printing them all at the end.. **However** this would be bad practice, since these errors could potentially never get printed if the functions times out or an exception occurs down the line. It's safer to log them as they occur in production.